src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java

```diff
@@ -777,47 +786,69 @@ public JCExpression parseExpression() {
 
     /** parses patterns.
      */
-
     public JCPattern parsePattern(int pos, JCModifiers mods, JCExpression parsedType,
                                   boolean allowVar, boolean checkGuard) {
         JCPattern pattern;
         mods = mods != null ? mods : optFinal(0);
         JCExpression e;
-        if (parsedType == null) {
-            boolean var = token.kind == IDENTIFIER && token.name() == names.var;
-            e = unannotatedType(allowVar, TYPE | NOLAMBDA);
-            if (var) {
-                e = null;
+        if (token.kind == UNDERSCORE && parsedType == null) {
+            nextToken();
+            pattern = toP(F.at(token.pos).AnyPattern());
+        }
+        else {
+            if (parsedType == null) {
+                boolean var = token.kind == IDENTIFIER && token.name() == names.var;
+                e = unannotatedType(allowVar, TYPE | NOLAMBDA);
+                if (var) {
+                    e = null;
+                }
+            } else {
+                e = parsedType;
             }
-        } else {
-            e = parsedType;
-        }
-        if (token.kind == LPAREN) {
-            //deconstruction pattern:
-            checkSourceLevel(Feature.RECORD_PATTERNS);
-            ListBuffer<JCPattern> nested = new ListBuffer<>();
-            if (!peekToken(RPAREN)) {
-                do {
+            if (token.kind == LPAREN) {
+                //deconstruction pattern:
+                checkSourceLevel(Feature.RECORD_PATTERNS);
+                ListBuffer<JCPattern> nested = new ListBuffer<>();
+                if (!peekToken(RPAREN)) {
+                    do {
+                        nextToken();
+                        JCPattern nestedPattern = parsePattern(token.pos, null, null, true, false);
+                        nested.append(nestedPattern);
+                    } while (token.kind == COMMA);
+                } else {
                     nextToken();
-                    JCPattern nestedPattern = parsePattern(token.pos, null, null, true, false);
-                    nested.append(nestedPattern);
-                } while (token.kind == COMMA);
+                }
+                accept(RPAREN);
+                pattern = toP(F.at(pos).RecordPattern(e, nested.toList()));
             } else {
-                nextToken();
-            }
-            accept(RPAREN);
-            pattern = toP(F.at(pos).RecordPattern(e, nested.toList()));
-        } else {
-            //type test pattern:
-            JCVariableDecl var = toP(F.at(token.pos).VarDef(mods, ident(), e, null));
-            if (e == null) {
-                var.startPos = pos;
+                //type test pattern:
+                UnnamedDetails result = identOrFlagUnderscore(mods);
+                JCVariableDecl var = toP(F.at(result.varPos()).VarDef(mods, result.name(), e, null));
```

Comment:
The behavior I observe is that `variableDeclarator` does not behave well when we have a `var` instead of a type. The former is well suited for variable declarations outside type patterns. But I can investigate further ..

Reply:
What about reusing `variableDeclaratorRest`? 

https://github.com/openjdk/jdk/pull/13528/commits/3d83fa14ae26981d5b6fc75faa0d3ba9336952b1